bin/p2-label/main.go

```diff
@@ -0,0 +1,143 @@
+
+	cmdShow       = kingpin.Command(CmdShow, "Show labels that apply to a particular entity (type, ID)")
+	showLabelType = cmdShow.Flag("labelType", "The type of label to adjust. Sometimes called the \"label tree\"").Short('t').Required().String()
+	showID        = cmdShow.Flag("id", "The ID of the entity to show labels for.").Short('i').Required().String()
+)
+
+const (
+	CmdApply = "apply"
+	CmdShow  = "show"
+)
+
+func main() {
+	cmd, opts := flags.ParseWithConsulOptions()
+	client := kp.NewConsulClient(opts)
+	applicator := labels.NewConsulApplicator(client, 3)
+
+	switch cmd {
+	case CmdShow:
+		labelType, err := labels.AsType(*showLabelType)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error while parsing label type. Check the commandline. \n%v\n", err)
+		}
+
+		labelsForEntity, err := applicator.GetLabels(labelType, *showID)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Got error while querying labels. %v\n", err)
+			return
+		}
+		fmt.Printf("The current labels for %s are: %s\n", *showID, labelsForEntity.Labels.String())
+		return
+	case CmdApply:
+		labelType, err := labels.AsType(*applyLabelType)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error while parsing label type. Check the commandline. \n%v\n", err)
+		}
+
+		subject, err := klabels.Parse(*applySubjectSelector)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error while parsing subject label. Check the syntax. \n%v\n", err)
+		}
+
+		additive, err := klabels.Parse(*applyAddititiveLabels)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error while parsing additive label set. Check the syntax. \n%v\n", err)
+		}
+		destructive, err := klabels.Parse(*applyDestructiveLabels)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error while parsing destructive label set. Check the syntax. \n%v\n", err)
+		}
+
+		cachedMatch := false
+		matches, err := applicator.GetMatches(subject, labelType, cachedMatch)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error while finding label matches. Syntax error? \n%v\n", err)
+		}
+
+		for _, match := range matches {
+			entityID := match.ID
+
+			applyLabels(applicator, entityID, labelType, additive, destructive)
+		}
+		return
+	}
+
+}
+
+func applyLabels(applicator labels.Applicator, entityID string, labelType labels.Type, additiveLabels, destructiveLabels klabels.Selector) error {
+	var err error
+
+	labelsForEntity, err := applicator.GetLabels(labelType, entityID)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Got error while querying labels. %v\n", err)
+		return err
+	}
+	fmt.Printf("The current labels for %s are: %s\n", entityID, labelsForEntity.Labels.String())
+	if !additiveLabels.Empty() {
+		fmt.Printf("labels to be added: %s\n", additiveLabels.String())
+	}
+
+	if !destructiveLabels.Empty() {
+		fmt.Printf("labels to be removed: %s\n", destructiveLabels.String())
+	}
+	fmt.Println("Continue?")
```

Comment:
Please confirm that you want to have an explicit confirmation for every host this applies to. It might be nicer to confirm once for the batch.